### ISSUE TITLE
Blocked version of Cholesky gradients

### ISSUE THREAD

**Issue by rtrangucci** [stan-dev/math#384]
#### Summary:

Update our Cholesky differentiation algorithm from Giles' to Murray's:

https://arxiv.org/pdf/1602.07527.pdf
#### Description:

Current algorithm doesn't use memory locality to its advantage, yielding a slower algorithm than is possible with blocking.
#### Current Version:

v2.11.0

**Comment by bob-carpenter** [comment]
That would be awesome.

**Comment by rtrangucci** [comment]
see here https://github.com/stan-dev/math/blob/a182f8cc7da99d6e74b9dc97347aaf7621ad99bc/stan/math/rev/mat/fun/cholesky_decompose.hpp

Any chance you could take a look and tell me if you see any obvious efficiency gains? I used this branch plus the new `cov_exp_quad` to estimate an N = 1280 GP in 35 mins last night on a compute-intensive AWS over the weekend, so this seems to be quite a bit faster than the current version. Not sure we'd even be able to estimate the N = 1280 GP with the current version.

**Comment by rtrangucci** [comment]
Forgot to add, I ran the N=1280 one-dim GP for 500 warmup iterations and 500 post-warmup iterations

**Comment by bgoodri** [comment]
Can we open an issue to recruit Iain Murray to contribute to Stan?

**Comment by rtrangucci** [comment]
I think that would be great, I was going to reach out to him anyway to see if he wanted to take a look at this branch.

**Comment by rtrangucci** [comment]
@betanalpha @avehtari 

N = 1280, AWS c4.2xlarge:

Ran 2 instances at @syclik's behest to see whether there were any timing differences between boxes, both runs are below. I generated 1280 data points from a GP where length scale = sqrt(3) (`length` in the inferences below) and amplitude = 0.5 (`alpha` in the inferences below). All chains were run from the same seed `383281`.

```
Inference for Stan model: stan_new_code_model
4 chains: each with iter=(500,500,500,500); warmup=(0,0,0,0); thin=(1,1,1,1); 2000 iterations saved.

Warmup took (1095, 1098, 1096, 1096) seconds, 1.2 hours total
Sampling took (1053, 1054, 1053, 1052) seconds, 1.2 hours total

                 Mean     MCSE   StdDev        5%    50%    95%  N_Eff  N_Eff/s    R_hat
lp__             2268  3.1e-02  1.0e+00   2.3e+03   2268   2269   1116  2.7e-01  1.0e+00
accept_stat__    0.87  3.8e-03  1.7e-01   4.8e-01   0.95    1.0   1947  4.6e-01  1.0e+00
stepsize__       0.61  7.8e-16  1.1e-15   6.1e-01   0.61   0.61    2.0  4.8e-04  1.0e+00
treedepth__       1.9  1.6e-02  6.2e-01   1.0e+00    2.0    3.0   1440  3.4e-01  1.0e+00
n_leapfrog__      5.2  4.9e-02  2.1e+00   3.0e+00    5.0    7.0   1788  4.2e-01  1.0e+00
divergent__      0.00  0.0e+00  0.0e+00   0.0e+00   0.00   0.00   2000  4.7e-01     -nan
energy__        -2267  5.2e-02  1.7e+00  -2.3e+03  -2267  -2264   1028  2.4e-01  1.0e+00
beta              1.5  1.4e-02  4.4e-01   8.1e-01    1.4    2.3   1081  2.6e-01  1.0e+00
length            1.5  5.0e-03  1.6e-01   1.2e+00    1.5    1.7   1057  2.5e-01  1.0e+00
delta           0.010  1.3e-05  3.8e-04   9.6e-03  0.010  0.011    906  2.2e-01  1.0e+00
alpha            0.98  6.4e-03  2.4e-01   6.2e-01   0.97    1.4   1346  3.2e-01  1.0e+00
```

```
4 chains: each with iter=(500,500,500,500); warmup=(0,0,0,0); thin=(1,1,1,1); 2000 iterations saved.

Warmup took (1082, 1087, 1087, 1087) seconds, 1.2 hours total
Sampling took (1051, 1050, 1050, 1050) seconds, 1.2 hours total

                 Mean     MCSE   StdDev        5%    50%    95%  N_Eff  N_Eff/s    R_hat
lp__             2268  3.1e-02  1.0e+00   2.3e+03   2268   2269   1116  2.7e-01  1.0e+00
accept_stat__    0.87  3.8e-03  1.7e-01   4.8e-01   0.95    1.0   1947  4.6e-01  1.0e+00
stepsize__       0.61  7.8e-16  1.1e-15   6.1e-01   0.61   0.61    2.0  4.8e-04  1.0e+00
treedepth__       1.9  1.6e-02  6.2e-01   1.0e+00    2.0    3.0   1440  3.4e-01  1.0e+00
n_leapfrog__      5.2  4.9e-02  2.1e+00   3.0e+00    5.0    7.0   1788  4.3e-01  1.0e+00
divergent__      0.00  0.0e+00  0.0e+00   0.0e+00   0.00   0.00   2000  4.8e-01     -nan
energy__        -2267  5.2e-02  1.7e+00  -2.3e+03  -2267  -2264   1028  2.4e-01  1.0e+00
beta              1.5  1.4e-02  4.4e-01   8.1e-01    1.4    2.3   1081  2.6e-01  1.0e+00
length            1.5  5.0e-03  1.6e-01   1.2e+00    1.5    1.7   1057  2.5e-01  1.0e+00
delta           0.010  1.3e-05  3.8e-04   9.6e-03  0.010  0.011    906  2.2e-01  1.0e+00
alpha            0.98  6.4e-03  2.4e-01   6.2e-01   0.97    1.4   1346  3.2e-01  1.0e+00

```

**Comment by rtrangucci** [comment]
Currently running N = 5120 right now for 150 warmup, 150 samples on each box

**Comment by betanalpha** [comment]
Maybe have two kinds of plots.

For each parameter plot the 5-95 interval minus the true value as a band vs N,
then overlay the median minus the true value as a line vs N.  This will verify
that for each N we’re recovering a reasonable posterior.

Then on a single graph plot Neff/s (or s/Neff) as a function of N for each
parameter to see the performance scaling.

**Comment by rtrangucci** [comment]
closed with #497
